content/blog/gsoc-25-netjsongraph.rst

```diff
@@ -0,0 +1,225 @@
+
+    .. code-block:: javascript
+
+       // file: src/js/netjsongraph.render.js
+       self.leaflet.on("zoomend", () => {
+         const currentZoom = self.leaflet.getZoom();
+         const showLabel = currentZoom >= self.config.showLabelsAtZoomLevel;
+         self.echarts.setOption({
+           series: [{ id: "geo-map", label: { show: showLabel } }]
+         });
+       });
+
+- **Graph Mode:** The solution uses the ECharts ``graphRoam`` event and a custom label ``formatter`` function.
+
+    .. code-block:: javascript
+
+       // file: src/js/netjsongraph.render.js (inside generateGraphOption)
+       if (typeof self.config.showGraphLabelsAtZoom === "number") {
+         const threshold = self.config.showGraphLabelsAtZoom;
+         baseGraphLabel.formatter = (params) =>
+           getGraphZoom() >= threshold ? params.data.name || "" : "";
+       }
+
+Resiliency and Robustness Fixes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- **Duplicate Node IDs:** I used a "defense-in-depth" strategy to prevent crashes from invalid data. The core logic is a new ``deduplicateNodesById`` utility.
+
+    .. code-block:: javascript
+
+       // file: src/js/netjsonWorker.js
+       deduplicateNodesById(nodes) {
+         const uniqueNodes = [];
+         const nodeIds = new Set();
+         nodes.forEach((node) => {
+           if (node.id) {
+             if (!nodeIds.has(node.id)) {
+               nodeIds.add(node.id);
+               uniqueNodes.push(node);
+             } else {
+               console.warn(`Duplicate node ID detected and skipped: ${node.id}`);
+             }
+           } else {
+             uniqueNodes.push(node); // Keep nodes without IDs
+           }
+         });
+         return uniqueNodes;
+       }
+
+- **Map Zoom Levels:** I corrected the map's behavior to respect tile provider limits and added CSS for visual feedback on the zoom controls.
+
+    .. code-block:: css
+
+       /* file: src/css/netjsongraph-theme.css */
+       .leaflet-control-zoom-in.leaflet-disabled,
+       .leaflet-control-zoom-out.leaflet-disabled {
+         opacity: 0.7;
+         cursor: default;
+       }
+
+Conclusion
+----------
+
+The Google Summer of Code program has been an incredibly rewarding experience. I successfully delivered all the required features, significantly enhancing the ``netjsongraph.js`` library's capabilities and user experience. The new clustering mechanism, in particular, is a major advancement that makes geographic network visualizations far more insightful. Through this project, I have deepened my skills in JavaScript, data visualization with ECharts and Leaflet, and robust software engineering practices like writing comprehensive tests and documentation. I am deeply grateful to my mentors and the entire OpenWISP community for their invaluable guidance and support throughout the summer.
```

Comment:
Here again link the repo. Try to break this paragraph into at least two apragraphs.

Mention your mentors by name and link their github profile please.